highfidelity/hifi-content
# Add latest sit script

**Test Plan for Sit Scripts:**

Setup:

1. Create sittable entity, can either:
     - Import from json: [Download Blue Chair JSON](https://hifi-production.s3.amazonaws.com/robin/sit/sitPrototypes/bluechairlatest.json)
     - Or add to an entity. Use the registration point to be the seat center. Ex - a stool would use a registration point of { x: 0.5, y: 1.0, z: 0.5 }
2. Ensure both the entity script and server script is running on Entity
3. Spawn multiple chairs. Best tested with multiple avatars and users see "Multiple chair/user" test section

**Desktop regular use case**
1. Walk within 3 m from the chair
2. “Click to Sit” overlay appears on chair flat on the seat and in the chair direction
3. Should fade to a barely visible dark grey over time
4. Click on the overlay
5. Pop into chair facing forward (without HMD overlay) and sit
6. Can swivel around in chair with mouse
7. Use any movement key and a “Hold to Stand” overlay will appear in front of view
8. Should pop in front of the chair facing the direction the avatar was facing

**HMD regular use case - differences from Desktop in bold**
1. Walk within 3 m from the chair
2. **“Trigger to Sit”** overlay appears on chair flat on the seat and in the chair direction
3. Should fade to a barely visible dark grey over time
4. Trigger on the overlay
5. **Overlay to Face Forward appears for one second for user to get comfortable in real space**
6. **Timed at the end of the overlay, Your avatar head and camera should be facing forward in the chair with your legs facing forward**
7. Can **snap** and rotate around in chair with **right joystick**
8. **Head can look around normally without popping out of the chair**
9. Use any movement key and a “Hold to Stand” overlay will appear in front of view
10. Should pop in front of the chair facing the direction the avatar was facing

**Overlay tests**
- Sittable overlay appears within 2.8 m of walking to the chair disappears after walking outside 5 m
- Face Forward and Hold to Stand overlay should stay in middle of view
- In HMD mode always see Trigger to Sit, in Desktop mode always see Click to Sit

**Get out of chair tests**
- In HMD, can move away from old position and pop out of chair
- Hold down movement keys including flying
- Disconnects from server then reloads it
- Pops out of chair when:
   - Resize avatar
   - Change Avatar
   - Change domain
   - Teleport out of chair

**Multiple chair/user tests**
- Overlay 3m/5m apply
- While sitting, can switch chairs using the overlay. Old chair become available for another person
- Click/trigger to sit overlay disappears shortly after another HMD user triggers to sit on chair

**Outlier cases to ensure do not happen**
- Avatar never has the sit animation active outside of a chair 
- 2 avatars can not sit in the same chair _- see Known Bug section_
- Should not see sittable overlay when another avatar is sitting in chair
- Chair does not lock up. If avatar disconnects suddenly, the heartbeat interval on the server script should allow chair to be settable again within about 11 seconds.

**Known Bugs - that happen <5% of the time**
- In HMD mode, after sit, avatar sometimes is not rotated forward in the chair
- 2 avatars sit in the same chair if they trigger/click at the exact same time
- Significant lag in a domain can cause an avatar in HMD to pop out of chair without sitting - can correct by trying again
